### ISSUE TITLE
Inliers for Less textured Images

### ISSUE THREAD

**Issue by engraamirkhan** [mihaidusmanu/d2-net#8]
Hi,

I am planning to use this architecture to find the feature points and subsequently inliers after matching for pair of images that consist of less texture.

The pre-trained model is returning zero inliers.

Does this mean, I have to perform the training with such images or do you have a better suggestion?

Thank you.

**Comment by mihaidusmanu** [comment]
Hello,

Finetuning on a similar dataset could definitely help, but I find it strange that there are no inliers - in our indoor evaluation scenario there are quite a few scenes with textureless areas, but D2 still finds some matchable points.

Could you provide me with an image pair from your dataset (if it is public, of course)? Also, how do you compute the number of inliers (RANSAC + geometric transformation fitting or ground truth transformation / correspondences)? Did you try increasing the inlier threshold?

In case you are using a ground truth transformation, make sure that it uses the same convention as our features (X points to the right, Y points down for pixel positions).

Regards,
Mihai.

**Comment by engraamirkhan** [comment]
Thank you for the prompt reply.
I used RANSAC from opencv to find the inliers, I tried both with the ratio test and without to find the good matches. 
I am attaching a pair of images that you may be able to test.
[image2.zip](https://github.com/mihaidusmanu/d2-net/files/3368700/image2.zip)


Thanks for you help.
Much appreciated.

**Comment by mihaidusmanu** [comment]
I would expect D2 to work without any issues on the image pair that you shared (there's a significant amount of texture / objects in the background that look matchable to me). I don't have time to look into it right now but I'll update you as soon as I do.

The only suggestion that I have right now is try different inlier thresholds. I have never used the OpenCV RANSAC with D2 features myself. The inlier threshold of the essential matrix estimation from OpenCV is in the normalized coordinates - see https://image-matching-workshop.github.io/slides/slides-matas.pdf (slides 17 and after). You might want to try something along the lines 2 / focal_length.

In the paper we either used the LORANSAC from COLMAP (https://github.com/colmap/colmap) or the one provided with the InLoc implementation (https://github.com/HajimeTaira/InLoc_demo).

Regarding matching, in stereo scenarios when you only want a relative pose, I'd suggest using mutual NN (without any kind of ratio test) or even one way NN.

**Comment by engraamirkhan** [comment]
Thank you very much on clarification and helpful links. I will try the suggested solutions and update with the results. 
We have this dataset captured by a UAV and the end goal is to have the 3D profile of the wind turbine blade. With most of the techniques, I have tried, finding meaningful (inliers) features on the blade is a difficult task. D2-Net appears to be a good fit to go about solving this problem. Hence, I appreciate the development of this work and making it public.
If you find some time to do the test on your own, please let me know your thoughts on the results.

Thanks again.

**Comment by mihaidusmanu** [comment]
Hello. I am sorry for the delay, I just checked the image pair you shared with me and I actually get a significant number of inliers (1.6k) using the Fundamental Matrix Estimation from scikit image (see https://scikit-image.org/docs/dev/auto_examples/transform/plot_fundamental_matrix.html) even with a low threshold (1px). This is using the 8 point algorithm and a very basic RANSAC implementation. Using the 5 point algorithm (with known intrinsics) and a better RANSAC will probably improve the results even further. 

You can check out the notebook for this at [Fundamental.zip](https://github.com/mihaidusmanu/d2-net/files/3411315/Fundamental.zip).

**Comment by engraamirkhan** [comment]
Hi,

Thank you so much for looking into it and explanation. 
I will be also training the model with megadepth dataset and will update you or ask if any issues arise.

Thanks again.

**Comment by mihaidusmanu** [comment]
Hello. We released a small update to the dataset generation that makes the training process more stable and the model now converges significantly faster. In case you are still planning on experimenting with MegaDepth, don't forget to pull the latest version. I will close this issue but feel free to open a new one in case you run into any problems / etc.

**Comment by engraamirkhan** [comment]
Hi,

Thank you so much.
